src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/Publishers.cs

```diff
@@ -12,8 +12,28 @@ public class Publishers
     {
         internal IEnumerable<PublisherAddress> GetPublisherFor(Type eventType)
         {
-            var distinctPublishers = rules.Select(r => r.Apply(eventType)).Where(e => e != null).Distinct().ToList();
-            return distinctPublishers;
+            List<PublisherAddress> staticPublishersForType;
+            if (!staticPublishers.TryGetValue(eventType, out staticPublishersForType))
+            {
+                staticPublishersForType = emptyList;
+            }
+
+            if (dynamicRules.Count > 0)
+            {
+                var dynamicAddresses = new List<PublisherAddress>();
+                foreach (var rule in dynamicRules)
+                {
+                    var address = rule(eventType);
+                    if (address != null)
+                    {
+                        dynamicAddresses.Add(address);
+                    }
+                }
+
+                return staticPublishersForType.Count > 0 ? staticPublishersForType.Concat(dynamicAddresses) : dynamicAddresses;
+            }
+
+            return staticPublishersForType;
```

Comment:
> Or, we can just keep it as is, and leave deduping up to the caller once they've resolved the addresses to strings!

keeping as is would definitely be the simpler approach but a proper Equals implementation would probably also save us some instance lookups which maybe worth the effort?